### ISSUE TITLE
ASan failure on test_bitcoin startup in latent0 (AWS).

### ISSUE THREAD

**Issue by nathan-at-least** [zcash/zcash#2049]
Example log: https://ci.z.cash/builders/zcashASan/builds/114/steps/full-test-suite.sh/logs/stdio

Split out from https://github.com/zcash/zcash/issues/2026

**Comment by str4d** [comment]
Opened https://svn.boost.org/trac/boost/ticket/12864 upstream.

**Comment by daira** [comment]
This upstream PR addresses some ASan issues: https://github.com/bitcoin/bitcoin/pull/9512

**Comment by zebambam** [comment]
Sorry @defuse, no need for analysis here :)

**Comment by zebambam** [comment]
Hi @defuse can you please check that this was fixed in the upgrade of boost? (see boost link above)

**Comment by defuse** [comment]
Summarizing the error posted to Boost's trac:

1. There are at least two threads running [`CScheduler::serviceQueue`](https://github.com/zcash/zcash/blob/8752c7aef5e0356c52862d5d43ed3a658d804891/src/scheduler.cpp#L23), the function responsible for running scheduled tasks from a list.
2. On one thread, the free occurs on [this line](https://github.com/zcash/zcash/blob/master/src/scheduler.cpp#L54), which frees the first task in the task list.
3. On the other thread, the use-after-free occurs on [this line](https://github.com/zcash/zcash/blob/8752c7aef5e0356c52862d5d43ed3a658d804891/src/scheduler.cpp#L44), which `wait_until`'s the first task in the list's scheduled execution time.

At the time of the error, we were using boost 1.62.0. Here's the source code from that version and a comment indicating where the use-after-free happened according to the stack trace:

```
        template <class Duration>
        cv_status
        wait_until(
                unique_lock<mutex>& lock,
                const chrono::time_point<chrono::system_clock, Duration>& t)
        {
          using namespace chrono;
          typedef time_point<system_clock, nanoseconds> nano_sys_tmpt;
          wait_until(lock,
                        nano_sys_tmpt(ceil<nanoseconds>(t.time_since_epoch())));
          return system_clock::now() < t ? cv_status::no_timeout : // <-- USE AFTER FREE HERE
                                             cv_status::timeout;
        }
```

`wait_until` takes its time argument by reference, unlocks the lock for a bit which gives the other thread an opportunity to free the element of the list holding the time, and then `wait_until` accesses the reference again.

In the current version of boost, [this problem is still there](https://github.com/boostorg/thread/blob/a645ef761ddf4c9811eb707c479363a79ae4534e/include/boost/thread/pthread/condition_variable_fwd.hpp#L265) although some of the other `wait_until` overloads make a copy of the time right away. The [docs](https://www.boost.org/doc/libs/1_69_0/doc/html/thread/synchronization.html#thread.synchronization.condvar_ref.condition_variable.wait_until) don't make any promises that the call should still work even if the time reference gets freed, so I think this is our bug.

**Comment by defuse** [comment]
Oh, this was already fixed in upstream: https://github.com/bitcoin/bitcoin/commit/12519bf62b8c49b1c1744eca6ea5b3162a61f962

**Comment by zebambam** [comment]
Ticket number #4569 is a dupe of this one :)

**Comment by daira** [comment]
This was fixed in #4600.
